### PR TITLE
Allow to run web app within Gnosis Safe

### PR DESCRIPTION
**Pull request by rmeissner**
This PR adds support to use the ENS app as a Gnosis Safe app within the Gnosis Safe web interface.

## List of features added/changed

- Use the https://www.npmjs.com/package/@gnosis.pm/safe-apps-sdk to check if we are connected to the webinterface
- Use https://www.npmjs.com/package/@gnosis.pm/safe-apps-provider to create a provider that can be used within the app
- Hide disconnect button when running as Safe app

## How Has This Been Tested?

- Follow steps outline in `docs/gnosis_safe_app_support.md` (part of this PR)

## Future improvements

- Proper icon for the Safe app (currently uses the Android one)
- Add padding to comensate for hiding the disconnect button

## Checklist:

- [ x] My code follows the code style of this project.
- [ x] My code implements all the required features.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,14 +8,15 @@
...
     "@emotion/styled": "^10.0.9",
-    "@myetherwallet/mewconnect-web-client": "^2.1.11",
-    "@portis/web3": "^2.0.0-beta.59",
-    "@toruslabs/torus-embed": "^1.8.6",
-    "@walletconnect/web3-provider": "^1.3.1",
     "@ensdomains/address-encoder": "^0.2.6",
     "@ensdomains/mock": "^2.0.36",
     "@ensdomains/react-ens-address": "^0.0.27",
     "@ensdomains/ui": "^3.0.63",
+    "@gnosis.pm/safe-apps-provider": "0.1.0",
```
Comment (rmeissner) [inline package.json:15]: this was an automatic reordering of the dependencies by yarn :/

File: package.json [hunk 2]
```diff
@@ -57,11 +58,11 @@
...
     "slugify": "^1.4.6",
     "stackdriver-errors-js": "^0.8.0",
-    "web3modal": "^1.9.1",
-    "web3": "^1.3.0"
+    "web3": "^1.3.0",
+    "web3modal": "^1.9.1"
   },
   "scripts": {
-    "start": "react-scripts start",
+    "start": "HTTPS=true react-app-rewired start",
```
Comment (rmeissner) [inline package.json:65]: I would remove this before making the PR final, as this is only required for local testing of the Safe apps

File: config-overrides.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
```
Comment (rmeissner) [inline config-overrides.js:1]: This files should be removed and maybe add documentation to the readme a link how to test Safe Apps

File: src/i18n.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ i18n
       escapeValue: false // not needed for react as it escapes by default
     },
     backend: {
-      loadPath: '/locales/{{lng}}.json'
+      loadPath: './locales/{{lng}}.json'
```
Comment (rmeissner) [inline src/i18n.js:22]: This was required to build the ipfs version (was used for testing) and should be reverted

File: public/manifest.json [hunk 5]
```diff
@@ -1,6 +1,8 @@
 {
   "short_name": "ENS App",
   "name": "ENS App",
+  "description": "ENS App",
```
Comment (rmeissner) [inline public/manifest.json:4]: Probably a little better description would also be nice :D

File: package.json (hunk 2 above)
Comment (makoto) [reply package.json:65]: Hi, so this PR is not ready to merge?

**Comment by makoto** [comment]
Hi, Thank you for your contribution. I see a couple of "should be removed/reverted" in your comment. Does this mean it's not ready yet? Can you also write down step by step guide of how I can test?

**Comment by rmeissner** [comment]
I think this was an open question from my side (not well formulated 😬 )
`package.json` and `config-overrides.js` these changes were required to test a local version of the app as a safe app. So you could keep the file so that you can easily test it as a Safe app at any time or you say that should rather be done when needed. In the later case I would remove it (as it is not required for the deployed version).

`src/i18n.js` here I was not sure if it was a mistake on my side so I would revert it, but I was wondering if that is expected

`public/manifest.json` I was just wondering if you have a preferred description that should be displayed in the Safe interface.

Also are there any guidelines for `yarn.lock`. I would probably regenerate it once to make sure there are no unnecessary changes of it.

Additionally to that there are the todos/ open question in the description:
- Potentially remove react-app-rewired and add to Readme how to test Safe app setup
  - These are mainly the changes to `package.json` and `config-overrides.js`
- Proper icon for the Safe app (currently uses the Android one)
- Add padding to compensate for hiding the disconnect button

If these todos don't need to be fixed or are ok in the current state I would get the PR ready for review (so that it is ready to be merged)

**Comment by makoto** [comment]
@jefflau can you comment on questions from @rmeissner  ?

**Comment by jefflau** [comment]
@makoto Have you managed to test the app at all?

@rmeissner Couple questions as I don't understand the premise of this PR as of yet.

Is this PR to support the ENS app within Gnosis safe? It's a bit confusing as I originally thought there was something unsafe with the app.

Can we do this PR without react-rewired as I don't know what that is and although it seems nice to add webpack functionality, I have no idea how stable this package is and I'm conservative about adding dependencies where it's unclear whether the author will continue to support it?

**Comment by rmeissner** [comment]
@jefflau 

Yes I can make this more clear in the title and description.  

We can definetly remove `react-rewired`, as mentioned before this is only required for testing it locally as a Gnosis Safe app (as browsers require the CORS headers which are normally not set by the dev servers). 

Should I add the instructions how to test this locally as a Gnosis Safe app somewhere?

**Comment by makoto** [comment]
> Should I add the instructions how to test this locally as a Gnosis Safe app somewhere?

yes, please.

**Comment by rmeissner** [comment]
@makoto @jefflau I removed `react-app-rewired` and related files. Also added a doc how to test it locally without changes to the repo configuration/ code.

File: docs/gnosis_safe_app_support.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
+## Running as a Gnosis Safe App
+
+It is possible to run the ENS app as a [Gnosis Safe App](https://docs.gnosis.io/safe/docs/sdks_safe_apps/). This will allow it to directly use the ENS app from withing the [Gnosis Safe web interface](https://app.gnosis-safe.io).
+
+As the ENS app is normally running on a different host (unless both, the Gnosis Safe web interface and the ENS app, are for example hosted locally) it is required to enable CORS headers. This will allow the Gnosis Safe web interface to access the ENS app meta information from the `manifest.json`.
```
Comment (makoto) [inline docs/gnosis_safe_app_support.md:5]: I think it's fine to make changes on ENS to allow CORS. Feel free to make the necessary changes.

**Comment by makoto** [comment]
I just tested on rinkeby and it looks like it registered the new name as a signer (`0x5a38`) than Gnosis multisig itself (`0x3f8`). Is this expected? 

Screenshot 2021-03-26 at 14 52 06

Can you also reach me out on Discord at ENS channel so that we can speed up the conversation? https://discord.gg/AskZbFx

**Comment by makoto** [comment]
Two outstanding items.

- Make it work when user refreshes the page.
- Must handle pending event properly

**Pull request merged**